test/val/val_interfaces_test.cpp

```diff
@@ -1939,6 +1939,321 @@ OpFunctionEnd
+     OpExecutionMode %main OriginUpperLeft
+     OpDecorate %cFlat Flat
+     OpDecorate %cFlat Location 0
+     OpDecorate %cFlat Component 0
+     OpDecorate %cFlat2 Flat
+     OpDecorate %cFlat2 Location 0
+     OpDecorate %cFlat2 Component 1
+     OpDecorate %cSmooth Centroid
+     OpDecorate %cSmooth Location 1
+     OpDecorate %cSmooth Component 0
+%void = OpTypeVoid
+%4 = OpTypeFunction %void
+%float = OpTypeFloat 32
+%in_ptr = OpTypePointer Input %float
+%cFlat = OpVariable %in_ptr Input
+%cFlat2 = OpVariable %in_ptr Input
+%cSmooth = OpVariable %in_ptr Input
+%main = OpFunction %void None %4
+%6 = OpLabel
+     OpReturn
+     OpFunctionEnd
+  )";
+
+  CompileSuccessfully(text, SPV_ENV_VULKAN_1_0);
+  EXPECT_EQ(SPV_SUCCESS, ValidateInstructions(SPV_ENV_VULKAN_1_0));
+}
+
+TEST_F(ValidateInterfacesTest, InterpolationOverlapVariableBad) {
+  const std::string text = R"(
+     OpCapability Shader
+     OpMemoryModel Logical GLSL450
+     OpEntryPoint Fragment %main "main" %cFlat %cSmooth
+     OpExecutionMode %main OriginUpperLeft
+     OpDecorate %cFlat Flat
+     OpDecorate %cFlat Location 0
+     OpDecorate %cFlat Component 0
+     OpDecorate %cSmooth Centroid
+     OpDecorate %cSmooth Location 0
+     OpDecorate %cSmooth Component 1
+%void = OpTypeVoid
+%4 = OpTypeFunction %void
+%float = OpTypeFloat 32
+%in_ptr = OpTypePointer Input %float
+%cFlat = OpVariable %in_ptr Input
+%cSmooth = OpVariable %in_ptr Input
+%main = OpFunction %void None %4
+%6 = OpLabel
+     OpReturn
+     OpFunctionEnd
+  )";
+
+  CompileSuccessfully(text, SPV_ENV_VULKAN_1_0);
+  EXPECT_EQ(SPV_ERROR_INVALID_DATA, ValidateInstructions(SPV_ENV_VULKAN_1_0));
+  EXPECT_THAT(getDiagnosticString(),
+              AnyVUID("VUID-StandaloneSpirv-Input-10604"));
+  EXPECT_THAT(
+      getDiagnosticString(),
+      HasSubstr("input Location 0 has conflicting Interpolation decorations"));
+}
+
+TEST_F(ValidateInterfacesTest, InterpolationOverlapBlockSuccess) {
+  const std::string text = R"(
+OpCapability Shader
+OpMemoryModel Logical GLSL450
+OpEntryPoint Fragment %main "main" %in
+OpExecutionMode %main OriginUpperLeft
+OpDecorate %struct Block
+OpMemberDecorate %struct 0 Flat
+OpMemberDecorate %struct 0 Location 0
+OpMemberDecorate %struct 0 Component 0
+OpMemberDecorate %struct 2 Flat
+OpMemberDecorate %struct 2 Location 0
+OpMemberDecorate %struct 2 Component 1
+OpMemberDecorate %struct 1 Centroid
+OpMemberDecorate %struct 1 Location 1
+OpMemberDecorate %struct 1 Component 0
+%void = OpTypeVoid
+%float = OpTypeFloat 32
+%struct = OpTypeStruct %float %float %float
+%in_ptr = OpTypePointer Input %struct
+%in = OpVariable %in_ptr Input
+%void_fn = OpTypeFunction %void
+%main = OpFunction %void None %void_fn
+%entry = OpLabel
+OpReturn
+OpFunctionEnd
+  )";
+
+  CompileSuccessfully(text, SPV_ENV_VULKAN_1_0);
+  EXPECT_EQ(SPV_SUCCESS, ValidateInstructions(SPV_ENV_VULKAN_1_0));
+}
+
+TEST_F(ValidateInterfacesTest, InterpolationOverlapBlockMixMembers) {
```

Comment:
We have https://gitlab.khronos.org/vulkan/vulkan/-/issues/4200 open

Things like this "might" be invalid, but went more conservative with the validation  in this PR because it is new, covering the "real" use case, and can always restrict more in the future (and we have tests already then!)